lib/hash.c

```diff
@@ -2253,17 +2255,24 @@ grn_io_hash_add(grn_ctx *ctx, grn_hash *hash, uint32_t hash_value,
     entry_id = header->curr_rec + 1;
     entry = grn_hash_entry_at(ctx, hash, entry_id, GRN_TABLE_ADD);
     if (!entry) {
+      ERR(GRN_UNKNOWN_ERROR, "[io][hash][add] failed to add entry(%u)", entry_id);
       return GRN_ID_NIL;
     }
     header->curr_rec = entry_id;
   }
 
   if (!grn_io_array_bit_on(ctx, hash->io, GRN_HASH_BITMAP_SEGMENT, entry_id)) {
     /* TODO: error handling. */
+    ERR(GRN_UNKNOWN_ERROR, "[io][hash][add] failed to handle bitmap segment about entry_id(%u)", entry_id);
   }
 
-  if (grn_hash_entry_put_key(ctx, hash, entry, hash_value, key, key_size)) {
-    grn_hash_delete_by_id(ctx, hash, entry_id, NULL);
+  rc = grn_hash_entry_put_key(ctx, hash, entry, hash_value, key, key_size);
+  if (rc != GRN_SUCCESS) {
+    ERR(rc, "[hash][put][key] failed to put key(%.*s)", key_size, (char *)key);
```

Comment:
Could you use common tag prefix in the same function?
`[io][hash][add]` for this function.